### PR TITLE
Learning rules updated, now modular and separate from Network object

### PR DESCRIPTION
**Pull request by BasBuller**

### CODE REVIEW

File: pysnn/utils.py [hunk 1]
```diff
@@ -47,3 +47,18 @@ def tensor_clamp(tensor, min, max):
...
+
+
+#########################################################
+# Select state dicts for set of modules
+#########################################################
+def make_layer(pre=None, connection=None, post=None):
+    # TODO: Move this into constructing an SNN network
+    states = {}
+    if pre:
+        states["pre_neuron"] = pre.state_dict()
```
Comment (Huizerd) [inline pysnn/utils.py]: State dicts consist of pointers?

File: pysnn/neuron.py [hunk 2]
```diff
@@ -62,6 +64,7 @@ def update_trace(self, x):
 
     def forward(self, x):
         self.update_trace(x)
+        self.spikes.copy_(x)
```
Comment (Huizerd) [inline pysnn/neuron.py]: Now, `Input` is suited to take either spikes or scalar values to pass on, which is nice and flexible. But maybe we should check in some way or differentiate between the two?

File: pysnn/learning.py [hunk 3]
```diff
@@ -8,58 +11,59 @@
...
-        self.reset_state()
+class LearningRule:
+    r"""Base class for correlation based learning rules in spiking neural networks.
+    
+    Arguments:
+        layers (iterable): an iterable or :class:`dict` of :class:`dict`s. 
+            The latter is a dict that contains a :class:`pysnn.Connection`s state dict, a pre-synaptic :class:`pysnn.Neuron`s state dict, 
+            and a post-synaptic :class:`pysnn.Neuron`s state dict that together form a single layer. These objects their state's will be 
+            used for optimizing weights.
+            During initialization of a learning rule that inherits from this class it is supposed to select only the paramters it needs
```
Comment (Huizerd) [inline pysnn/learning.py]: Typo

File: pysnn/learning.py [hunk 4]
```diff
@@ -8,58 +11,59 @@
...
+            from these objects.
+            The higher lever iterable or :class:`dict` contain groups that use the same parameter during training. This is analogous to
+            PyTorch optimizers' parameter groups.
+        defaults: A dict containing default hyper paramters. This is a placeholder for possible changes later on, these groups would work
```
Comment (Huizerd) [inline pysnn/learning.py]: Typo

File: pysnn/learning.py [hunk 5]
```diff
@@ -1,3 +1,6 @@
+from collections import defaultdict
```
Comment (Huizerd) [inline pysnn/learning.py]: `defaultdict` is unused? So are `torch.nn`, `pysnn.connection` and `pysnn.utils`

File: pysnn/learning.py [hunk 6]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+
+        # Set default hyper parameters
+        self.lr = lr
+        self.w_init = w_init
+        self.a = a
+
+        # To possibly later support groups, without chaning interface
+        defaults = {"lr": lr, "w_init": w_init, "a": a}
+
+        # Select only necessary paramters
```
Comment (Huizerd) [inline pysnn/learning.py]: Typo

File: pysnn/learning.py [hunk 7]
```diff
@@ -8,58 +11,59 @@
...
-    dimension before weight update."""
+    def update_state(self):
+        r"""Update state parameters of LearningRule based on latest network forward pass."""
+        raise NotImplementedError
 
-    def __init__(self, connection, lr, w_init, a):
-        super(FedeSTDP, self).__init__(connection, lr)
-        self.register_buffer("w_init", torch.tensor(w_init, dtype=torch.float))
-        self.register_buffer("a", torch.tensor(a, dtype=torch.float))
+    def step(self):
```
Comment (Huizerd) [inline pysnn/learning.py]: `step` to indicate that it's not actually "forward"?

File: pysnn/learning.py [hunk 8]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+        for layer in self.layers:
+            w = layer["weight"]
+            trace = layer["trace"].view(-1, *w.shape)
+
+            # LTP and LTD
+            dw = w - self.w_init
+
+            # LTP computation
+            ltp_w = torch.exp(-dw)
+            ltp_t = torch.exp(trace) - self.a
```
Comment (Huizerd) [inline pysnn/learning.py]: Trace should be normalized to [0, 1] by scaling with the current maximum at the moment of firing

File: pysnn/learning.py [hunk 9]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+    dimension before weight update."""
+
+    def __init__(self, layers, lr, w_init, a):
+        # Check layer formats
+        self.check_layers(layers)
+
+        # Set default hyper parameters
+        self.lr = lr
+        self.w_init = w_init
+        self.a = a
```
Comment (Huizerd) [inline pysnn/learning.py]: `lr` should be > 0 and `a` should be < 1 for stability (see paper), so maybe check for this? Or at least warn?

File: pysnn/learning.py [hunk 10]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+            dw = w - self.w_init
+
+            # LTP computation
+            ltp_w = torch.exp(-dw)
+            ltp_t = torch.exp(trace) - self.a
+            ltp = ltp_w * ltp_t
+
+            # LTD computation
+            ltd_w = -(torch.exp(dw))
+            ltd_t = torch.exp(1 - trace) - self.a
```
Comment (Huizerd) [inline pysnn/learning.py]: Same normalization here for trace

File: pysnn/learning.py [hunk 11]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+            ltd = ltd_w * ltd_t
+
+            # Perform weight update
+            layer["weight"] += self.lr * (ltp + ltd).mean(0)
```
Comment (Huizerd) [inline pysnn/learning.py]: Is this the mean over batches?

File: pysnn/learning.py [hunk 12]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+
+
+#########################################################
+# Fede STDP
+#########################################################
+class FedeSTDP(LearningRule):
+    r"""STDP version for Paredes Valles, performs mean operation over the batch 
+    dimension before weight update."""
+
+    def __init__(self, layers, lr, w_init, a):
```
Comment (Huizerd) [inline pysnn/learning.py]: Why not give defaults here like with `MSTDPET`?

File: pysnn/learning.py [hunk 13]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+        # Check layer formats
+        self.check_layers(layers)
+
+        # Set default hyper parameters
+        self.lr = lr
+        self.w_init = w_init
+        self.a = a
+
+        # To possibly later support groups, without chaning interface
```
Comment (Huizerd) [inline pysnn/learning.py]: Typo

File: pysnn/learning.py [hunk 14]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+            new_layer["weight"] = layer["connection"]["weight"]
+            new_layer["e_trace"] = torch.zeros_like(layer["connection"]["weight"])
+            layers[idx] = new_layer
+
+        self.a_pre = a_pre
+        self.a_post = a_post
+        self.lr = lr
+        self.e_trace_decay = e_trace_decay
+
+        # To possibly later support groups, without chaning interface
```
Comment (Huizerd) [inline pysnn/learning.py]: Typo

File: pysnn/learning.py [hunk 15]
```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
...
+    def reset_state(self):
+        for layer in self.layers:
+            layer["e_trace"].fill_(0)
 
-    def forward(self, reward):
+    def step(self, reward):
         # TODO: add weight clamping?
-        self.connection.weight += self.lr * reward * self.e_trace
+        for layer in self.layers:
+            layer["weight"] += self.defaults["lr"] * reward * layer["e_trace"]
```
Comment (Huizerd) [inline pysnn/learning.py]: Use `self.lr`, right?

File: examples/mstdpet_example.py [hunk 16]
```diff
@@ -47,22 +48,19 @@ def __init__(self):
         super(SNN, self).__init__()
 
         # One layer
-        self.pre_neuron = LIFNeuronTraceLinear((batch_size, 1, inputs), *n_dynamics)
-        self.post_neuron = LIFNeuronTraceLinear((batch_size, 1, outputs), *n2_dynamics)
+        self.pre_neuron = LIFNeuron((batch_size, 1, inputs), *n_dynamics)
+        self.post_neuron = LIFNeuron((batch_size, 1, outputs), *n2_dynamics)
         self.linear = Linear(*c_shape, *c_dynamics)
-        self.rule = MSTDPET(self.linear, self.pre_neuron, self.post_neuron, *l_params)
+        # self.rule = MSTDPET(self.linear, self.pre_neuron, self.post_neuron, *l_params)
```
Comment (Huizerd) [inline examples/mstdpet_example.py]: Remove the comments?

File: pysnn/learning.py (hunk 7 above)
Comment (BasBuller) [reply pysnn/learning.py]: Two reasons, to differentiate it from a Module, but also since this is the naming convention in regular PyTorch.
